Launch renderer process via forking at timeout

joonghunpark opened the pull request:
This CL for Castanets to launch renderer process if launching via tcp connection is not established within timeout.

joonghunpark commented:
Currently, this CL does forking at timeout(10 seconds), but can't detect when tcp connection is established. We have to cancel the timeout manager's timer to cancel forking if launching via tcp connection has been established in this CL. To do that, we have to find the place where we can ensure that tcp connection has succeeded and signal to the timeout manager to cancel forking.

hyowon reviewed (changes requested):
Making the invitation dispatcher static and avoiding deletion will cause too many side effects.

joonghunpark commented:
I think we could add a function which is executed when tcp launching succeeds or timeout occurs,
in which it deletes the invitation dispatcher for clean-up. Then 'static' will not be problematic I think, and for this purpose, we have to choose where is the proper location to signal child process launching success via tcp connection.

hyowon commented:
Each time you launch a child process, a new invitation will be generated. I wonder how you can handle invitation dispatchers properly. I will review it again with the next patch.

joonghunpark commented:
In class ChildConnection::IOThreadContext,
void InitializeOnIOThread(
...
    if (connector_) {
      connector_->StartService(child_identity, std::move(service),
                               std::move(pid_receiver_request));
      connector_->BindInterface(child_identity, &child_);
    }
  }

connector_->StartService above starts renderer service with process and ChildConnection is made only once per RenderProcessHostImpl.

That's why we can't make renderer process when we send OutgoingInvitation second time.

For now, I revised the CL's behavior as below.

[WIP] This CL changes child process launching behavior as below.

1) If timeout occurs waiting child process to be launched,
then uses child process launched with 127.0.0.1 connected via tcp.
2) After the first tcp connected child process,
uses --enable-forking to launch child processes to use unix domain socket.


If needed, I'd like to investigate sad tab reload function for reference.

joonghunpark committed:
Launch renderer process via forking.

If launching via tcp connection is not established
within timeout(5 sec), use forked renderer process.

And if tcp connection is established, use that connection
and doesn't try to connect with unix domain socket.

hyowon commented:
See content/child/child_thread_impl.cc:511

```
if (!base::CommandLine::ForCurrentProcess()->HasSwitch(
	switches::kRendererClientId)) {
  base::CommandLine::ForCurrentProcess()->AppendSwitchASCII(
	  switches::kRendererClientId, std::to_string(1)); // workaround
  invitation = InitializeMojoIPCChannelTCP();
} else
  invitation = InitializeMojoIPCChannel();
```

hyowon commented:
- Fork a new child process after 3 seconds sleep.
- Retry invitation with the new process and unix domain socketpair information.
- Discard existing tcp sockets and reset to new sockets in channel_posix.

We have confirmed that the above steps work well.
Refer commit below.
https://github.com/hyowon/Castanets/commit/ed654bbbdf77d1881c1d3002f047eac437847b39

joonghunpark commented:
* CL Status
Currently, if launching via tcp connection is not established
within timeout(3 sec), use forked renderer process.

This CL needs to add branch not to fork if tcp connection was established.

joonghunpark commented:
Could you please review this change?
Now the CL works as expected.

hyowon commented on mojo/core/node_controller.cc:
```diff
@@ -508,6 +524,7 @@ void NodeController::SendBrokerClientInvitationOnIOThread(
   // called below.
 
   pending_invitations_.insert(std::make_pair(temporary_node_name, channel));
+  temp_node_name_ = temporary_node_name;
```
We can access a node channel without this variable. See broker_castanets.h:126.

joonghunpark replied:
Done.

hyowon commented on mojo/core/node_controller.cc:
```diff
@@ -516,6 +533,41 @@ void NodeController::SendBrokerClientInvitationOnIOThread(
+  CHECK(channel_ok);
+#endif  // defined(OS_WIN)
+
+  auto it = pending_invitations_.find(temp_node_name_);
```
BrokerCastanets already has a node_channel pointer. See broker_castanets.h:126

joonghunpark replied:
Done. I added a getter for it.

hyowon commented on mojo/core/node_controller.h:
```diff
@@ -277,6 +286,7 @@ class MOJO_SYSTEM_IMPL_EXPORT NodeController : public ports::NodeDelegate,
   Core* const core_;
   const ports::NodeName name_;
   const std::unique_ptr<ports::Node> node_;
+  base::RepeatingCallback<void()> tcp_success_callback_;
```
I think it is not appropriate for NodeController to have this callback function as a member variable. And why is it a RepeatingCallback type? Wouldn't it be better OnceCallback?

joonghunpark replied:
OnceCallback::Run() may only be invoked on a non-const rvalue, i.e. std::move(callback).Run().
After std::move the member tcp_success_callback_ will be nullptr, so it could be a little odd.

I declared base::RepeatingCallback<void()> tcp_success_callback_; at node_controller
to use it in NodeController::OnAcceptInvitation.

hyowon replied:
RepeatingCallback allows calling Run() any number of time. However, this callback will have to run only once per one invitation, if not, we should think there is a bug.

```
SendInvitation...(..., OnceCallback callback) {
    connection_callback_ = std::move(callback);
}

AcceptInvitation...() {
    DCHECK(!connection_callback_.is_null());
    std::move(connection_callback_).Run();
}
```

If you still think it looks odd, let's leave it without change. This is just my suggestion.

hyowon commented on mojo/core/node_controller.h:
```diff
@@ -344,6 +354,8 @@ class MOJO_SYSTEM_IMPL_EXPORT NodeController : public ports::NodeDelegate,
   // Channels to invitees during handshake.
   NodeMap pending_invitations_;
 
+  ports::NodeName temp_node_name_;
```
Same as above comment.

joonghunpark replied:
Done.

hyowon commented on content/browser/child_process_launcher_helper.cc:
```diff
@@ -76,12 +79,30 @@ ChildProcessLauncherHelper::ChildProcessLauncherHelper(
+
+void ChildProcessLauncherHelper::OnCastanetsRendererLaunchedViaTcp() {
+  tcp_connected_ = true;
+  success_or_timeout_event_.Signal();
```
relaunch_renderer_process_monitor_timeout_.Stop();

joonghunpark replied:
Done.

hyowon commented on mojo/core/broker_castanets.h:
```diff
@@ -127,6 +130,8 @@ class BrokerCastanets : public Channel::Delegate, public base::SyncDelegate {
 
   std::unique_ptr<CastanetsFenceQueue> fence_queue_;
 
+  base::OnceCallback<void()> tcp_success_callback_;
```
Both Nodecontroller and BrokerCastanets has a callback for TCP connection success?

joonghunpark replied:
This is just a residue from my previous patchset.
Let me remove this.

hyowon commented on mojo/core/channel_posix.cc:
```diff
@@ -116,6 +116,21 @@ class ChannelPosix : public Channel,
+
+  void ClearOutgoingMessages() override {
+    base::circular_deque<MessageView> messages;
+    std::swap(outgoing_messages_, messages);
```
Isn't it just okay to call outgoing_messages_.clear()?

joonghunpark replied:
Ok, I will change this as you pointed out.

joonghunpark reviewed (commented):
I removed temp_node_name_ and uses BrokerCastanets to retrieve node channel as you commented.
Could you please review change again?

hyowon commented on mojo/core/broker_castanets.h:
```diff
@@ -40,11 +40,15 @@ class BrokerCastanets : public Channel::Delegate, public base::SyncDelegate {
   ~BrokerCastanets() override;
 
+  scoped_refptr<NodeChannel> GetNodeChannel() { return node_channel_; }
+
```
Wouldn't it be better to add ResetNodeChannel() method rather than using this getter?

joonghunpark replied:
Ok, I will add ResetNodeChannel() also.

hyowon commented on mojo/core/node_controller.cc:
```diff
@@ -516,6 +532,41 @@ void NodeController::SendBrokerClientInvitationOnIOThread(
+  if (it == broker_hosts_.end())
+    return;
+  auto channel = it->second->GetNodeChannel();
+  channel->SetSocket(std::move(node_connection_params));
```
As I mentioned above, if we have ResetNodeChannel() method, these lines could be simpler.

joonghunpark replied:
I will apply this in the following patchset.

hyowon commented on content/browser/child_process_launcher_helper.cc:
```diff
@@ -22,6 +24,8 @@
 #include "content/browser/android/launcher_thread.h"
 #endif
 
+int kTcpConnectionTimeout = 3;
```
const and how about 5 sec as a default. My hands are too slow.

joonghunpark replied:
Let me change to 5 sec.

joonghunpark replied:
When I tried with OnceCallback, Mostly ok but sometimes the logs like below was observed.
Used LOG is,
void NodeController::OnAcceptInvitation() {
  LOG(INFO) << "tcp_success_callback_.is_null(): " << tcp_success_callback_.is_null();
}

I think the reason is some web pages including multiple cross domains uses multiple renderer process for each of these for security reason.
And for this reason, I think RepeatingCallback could be reasonable.

For your reference, I will leave the patch's url with OnceCallback impl.
https://github.com/joonghunpark/Castanets/commits/oncecallback

[6117:6154:0911/113107.155153:INFO:child_process_launcher_helper.cc(183)] PostLaunchOnLauncherThread
[6117:6117:0911/113107.155139:INFO:render_process_host_impl.cc(3914)] SKIP!!!!! RenderProcessHostImpl::CreateSharedRendererHistogramAllocator
[6117:6154:0911/113107.155253:INFO:child_process_launcher_helper.cc(213)] PostLaunchOnLauncherThread - mojo::OutgoingInvitation::Send
[6117:6120:0911/113107.156187:INFO:tcp_platform_handle_utils_posix.cc(134)] Server Sock fd for port 37149:498
[6117:6117:0911/113107.165928:INFO:render_process_host_impl.cc(3914)] SKIP!!!!! RenderProcessHostImpl::CreateSharedRendererHistogramAllocator
[6191:6191:0911/113107.618523:WARNING:chrome_main_delegate.cc(280)] process type 'renderer' should be created through the zygote.
[6191:6191:0911/113107.619544:INFO:broker_castanets.cc(147)] Connection Success: Unix Domain Socket
**[6117:6120:0911/113107.619648:INFO:node_controller.cc(971)] tcp_success_callback_.is_null(): 0**
[6192:6192:0911/113107.624037:WARNING:chrome_main_delegate.cc(280)] process type 'renderer' should be created through the zygote.
[6192:6192:0911/113107.625040:INFO:broker_castanets.cc(147)] Connection Success: Unix Domain Socket
**[6117:6120:0911/113107.625115:INFO:node_controller.cc(971)] tcp_success_callback_.is_null(): 1**

hyowon replied:
> When I tried with OnceCallback, Mostly ok but sometimes the logs like below was observed.
> Used LOG is,
> void NodeController::OnAcceptInvitation() {
> LOG(INFO) << "tcp_success_callback_.is_null(): " << tcp_success_callback_.is_null();
> }
> 
> I think the reason is some web pages including multiple cross domains uses multiple renderer process for each of these for security reason.

Seriously? If it is true, Castanets won't work at all. Please share the test site.

> And for this reason, I think RepeatingCallback could be reasonable.

multiple renderer processes -> multiple child process launchers -> multiple invitations -> multiple callbacks BUT that doesn't mean one callback runs multiple times. As I explained, a success callback has to run only once per invitation. Do you agree with this part?

If a success callback is null in AcceptInvitation(), this may be one of the two cases below.
1) you haven't set a proper callback in SendInvitation() -> bug
2) you try to run the same callback for the second time even though the callback had been already run. (The second connection success?) -> bug

> For your reference, I will leave the patch's url with OnceCallback impl.
> https://github.com/joonghunpark/Castanets/commits/oncecallback

Let's just leave this PR with a repeating callback. If necessary, I will create a PR.

joonghunpark replied:
>Seriously? If it is true, Castanets won't work at all. Please share the test site.

The nullptr OnceCallback occurs only on 2nd Tab. 3rd or 4th is ok it seems.
So I think my assumption about the issue was incorrect. Sorry for my immature assumption.

>multiple renderer processes -> multiple child process launchers -> multiple invitations -> >multiple callbacks BUT that doesn't mean one callback runs multiple times. As I >explained, a success callback has to run only once per invitation. Do you agree with this >part?

I agree with this.

>If a success callback is null in AcceptInvitation(), this may be one of the two cases below.

>1. you haven't set a proper callback in SendInvitation() -> bug
>2. you try to run the same callback for the second time even though the callback had >been already run. (The second connection success?) -> bug

I think this corresponds the case 2.

>Let's just leave this PR with a repeating callback. If necessary, I will create a PR.

 I see. Thank you for your review:)

joonghunpark commented:
I applied the comments in the previous patchset. Could you please review this change?

hyowon commented on content/browser/child_process_launcher_helper.cc:
```diff
@@ -22,6 +24,10 @@
 #endif
 
+#if defined(CASTANETS)
+int kTcpConnectionTimeout = 5;
```
add include here.

joonghunpark replied:
Done.

hyowon commented on mojo/core/broker_castanets.h:
```diff
@@ -77,9 +81,9 @@ class BrokerCastanets : public Channel::Delegate, public base::SyncDelegate {
-             CastanetsFenceManager* fence_manager);
+                  ConnectionParams connection_params,
+                  const ProcessErrorCallback& process_error_callback,
+                  CastanetsFenceManager* fence_manager);
```
I've already applied the same code cleanup. Don't put code for cleanup in this pr. One commit = one issue.

joonghunpark replied:
I checked your commit below, and this change made by 'git cl format' was not applied in the commit below.
This style change was made automatically by script(although I didn't intent to do that), so I guess it's ok to go.

d5e7ff134cf743a39c7ba11618afb20c373fc104
Author: Hyowon Kim <hw1008.kim@samsung.com>
Date:   Wed Sep 11 10:07:14 2019 +0900

    Code cleanup in BrokerCastanets.

joonghunpark reviewed (commented):
All the comments left in the previous patchset has been applied.
